File: code/rust/ch03/src/vm/instruction.rs
use super::opcodes::OPCODES;

const MAXARG_BX: isize = (1 << 18) - 1; // 262143
const MAXARG_SBX: isize = MAXARG_BX >> 1; // 131071

/*
 31       22       13       5    0
  +-------+^------+-^-----+-^-----
  |b=9bits |c=9bits |a=8bits|op=6|
  +-------+^------+-^-----+-^-----
  |    bx=18bits    |a=8bits|op=6|
  +-------+^------+-^-----+-^-----
  |   sbx=18bits    |a=8bits|op=6|
  +-------+^------+-^-----+-^-----
  |    ax=26bits            |op=6|
  +-------+^------+-^-----+-^-----
 31      23      15       7      0
*/
pub trait Instruction {
    fn opname(self) -> &'static str;
    fn opmode(self) -> u8;
    fn b_mode(self) -> u8;
    fn c_mode(self) -> u8;
    fn opcode(self) -> u8;
    fn abc(self) -> (isize, isize, isize);
    fn a_bx(self) -> (isize, isize);
    fn a_sbx(self) -> (isize, isize);
    fn ax(self) -> isize;
}

impl Instruction for u32 {
    fn opname(self) -> &'static str {
        OPCODES[self.opcode() as usize].name
    }

    fn opmode(self) -> u8 {
        OPCODES[self.opcode() as usize].opmode
    }

    fn b_mode(self) -> u8 {
        OPCODES[self.opcode() as usize].bmode
    }

    fn c_mode(self) -> u8 {
        OPCODES[self.opcode() as usize].cmode
    }

    fn opcode(self) -> u8 {
        self as u8 & 0x3F
    }

    fn abc(self) -> (isize, isize, isize) {
        let a = (self >> 6 & 0xFF) as isize;
        let c = (self >> 14 & 0x1FF) as isize;
        let b = (self >> 23 & 0x1FF) as isize;
        (a, b, c)
    }

    fn a_bx(self) -> (isize, isize) {
        let a = (self >> 6 & 0xFF) as isize;
        let bx = (self >> 14) as isize;
        (a, bx)
    }

    fn a_sbx(self) -> (isize, isize) {
        let (a, bx) = self.a_bx();
        (a, bx - MAXARG_SBX)
    }

    fn ax(self) -> isize {
        (self >> 6) as isize
    }
}

File: code/csharp/ch10/luavm/vm/Instruction.cs
using System;
using luavm.api;
using LuaVM = luavm.state.LuaState;

namespace luavm.vm
{
    public class Instruction
    {
        public Instruction(uint b)
        {
            self = b;
        }

        private readonly uint self;

        public int Opcode()
        {
            return (int) (self & 0x3F);
        }

        public (int, int, int) ABC()
        {
            return (
                (int) (self >> 6) & 0xFF,
                (int) (self >> 23) & 0x1FF,
                (int) (self >> 14) & 0x1FF
            );
        }

        public (int, int) ABx()
        {
            return ((int) ((self >> 6) & 0xFF),
                (int) (self >> 14));
        }

        private const int MAXARG_Bx = (1 << 18) - 1;
        private const int MAXARG_sBx = MAXARG_Bx >> 1;

        public (int, int) AsBx()
        {
            var tuple = ABx();
            return (tuple.Item1, tuple.Item2 - MAXARG_sBx);
        }

        public int Ax()
        {
            return (int) (self >> 6);
        }

        private string OpName()
        {
            return OpCodes.opcodes[Opcode()].name;
        }

        public byte OpMode()
        {
            return OpCodes.opcodes[Opcode()].opMode;
        }

        public byte BMode()
        {
            return OpCodes.opcodes[Opcode()].argBMode;
        }

        public byte CMode()
        {
            return OpCodes.opcodes[Opcode()].argCMode;
        }

        private static void printStack(LuaState ls)
        {
            var top = ls.GetTop();
            for (var i = 1; i <= top; i++)
            {
                var t = ls.Type(i);
                switch (t)
                {
                    case Consts.LUA_TBOOLEAN:
                        Console.Write($"[{ls.ToBoolean(i)}]");
                        break;
                    case Consts.LUA_TNUMBER:
                        Console.Write($"[{ls.ToNumber(i)}]");
                        break;
                    case Consts.LUA_TSTRING:
                        Console.Write($"[{ls.ToString(i)}]");
                        break;
                    default:
                        Console.Write($"[{ls.TypeName(t)}]");
                        break;
                }
            }

            Console.WriteLine();
        }

        public void Execute(ref LuaVM vm)
        {
            var action = OpCodes.opcodes[Opcode()].action;

            //Console.Write(Opcode());
            //printStack(vm);

//            var ints = new int[] {11, 45, 43, 4};
//            foreach (var i in ints)
//            {
//                if (Opcode() == i)
//                {
//                    Console.Write(Opcode());
//                    printStack(vm);
//                }
//            }

            if (action != null)
            {
                action(this, ref vm);
            }
            else
            {
                throw new Exception(OpName());
            }
        }
    }
}